### ISSUE TITLE
"Max Hover Light Count (2)" Exceeded Warning when playing Scene with Default HL2 profile

### ISSUE THREAD

**Issue by julenka** [microsoft/MixedRealityToolkit-Unity#8543]
## Describe the bug

1. Import MRTK 2.5
2. Configure blank scene with MRTK Default HoloLens 2 profile
3. Hit play
4. Bring up hand

Observer following warning in console window:

```
Max hover light count (2) exceeded.
UnityEngine.Debug:LogWarningFormat(String, Object[])
Microsoft.MixedReality.Toolkit.Utilities.HoverLight:AddHoverLight(HoverLight) (at Library/PackageCache/com.microsoft.mixedreality.toolkit.foundation@2.5.0-preview.20200910.3/Core/Utilities/StandardShader/HoverLight.cs:103)
Microsoft.MixedReality.Toolkit.Utilities.HoverLight:OnEnable() (at Library/PackageCache/com.microsoft.mixedreality.toolkit.foundation@2.5.0-preview.20200910.3/Core/Utilities/StandardShader/HoverLight.cs:55)
UnityEngine.Object:Instantiate(GameObject, Transform)
Microsoft.MixedReality.Toolkit.Input.BaseControllerPointer:SetCursor(GameObject) (at Library/PackageCache/com.microsoft.mixedreality.toolkit.foundation@2.5.0-preview.20200910.3/SDK/Features/UX/Scripts/Pointers/BaseControllerPointer.cs:84)
Microsoft.MixedReality.Toolkit.Input.BaseControllerPointer:OnEnable() (at Library/PackageCache/com.microsoft.mixedreality.toolkit.foundation@2.5.0-preview.20200910.3/SDK/Features/UX/Scripts/Pointers/BaseControllerPointer.cs:150)
Microsoft.MixedReality.Toolkit.Input.LinePointer:OnEnable() (at Library/PackageCache/com.microsoft.mixedreality.toolkit.foundation@2.5.0-preview.20200910.3/SDK/Features/UX/Scripts/Pointers/LinePointer.cs:93)
Microsoft.MixedReality.Toolkit.Teleport.TeleportPointer:OnEnable() (at Library/PackageCache/com.microsoft.mixedreality.toolkit.foundation@2.5.0-preview.20200910.3/SDK/Features/UX/Scripts/Pointers/TeleportPointer.cs:136)
Microsoft.MixedReality.Toolkit.Teleport.ParabolicTeleportPointer:OnEnable() (at Library/PackageCache/com.microsoft.mixedreality.toolkit.foundation@2.5.0-preview.20200910.3/SDK/Features/UX/Scripts/Pointers/ParabolicTeleportPointer.cs:33)
UnityEngine.Object:Instantiate(GameObject)
Microsoft.MixedReality.Toolkit.Input.BaseInputDeviceManager:CreatePointer(PointerOption&) (at Library/PackageCache/com.microsoft.mixedreality.toolkit.foundation@2.5.0-preview.20200910.3/Core/Providers/BaseInputDeviceManager.cs:303)
Microsoft.MixedReality.Toolkit.Input.BaseInputDeviceManager:RequestPointers(SupportedControllerType, Handedness) (at Library/PackageCache/com.microsoft.mixedreality.toolkit.foundation@2.5.0-preview.20200910.3/Core/Providers/BaseInputDeviceManager.cs:223)
Microsoft.MixedReality.Toolkit.Input.BaseInputSimulationService:GetOrAddControllerDevice(Handedness, ControllerSimulationMode) (at Library/PackageCache/com.microsoft.mixedreality.toolkit.foundation@2.5.0-preview.20200910.3/Core/Providers/InputSimulation/BaseInputSimulationService.cs:139)
Microsoft.MixedReality.Toolkit.Input.BaseInputSimulationService:UpdateControllerDevice(ControllerSimulationMode, Handedness, Object) (at Library/PackageCache/com.microsoft.mixedreality.toolkit.foundation@2.5.0-preview.20200910.3/Core/Providers/InputSimulation/BaseInputSimulationService.cs:78)
Microsoft.MixedReality.Toolkit.Input.InputSimulationService:LateUpdate() (at Library/PackageCache/com.microsoft.mixedreality.toolkit.foundation@2.5.0-preview.20200910.3/Core/Providers/InputSimulation/InputSimulationService.cs:385)
Microsoft.MixedReality.Toolkit.BaseDataProviderAccessCoreSystem:LateUpdate() (at Library/PackageCache/com.microsoft.mixedreality.toolkit.foundation@2.5.0-preview.20200910.3/Core/Services/BaseDataProviderAccessCoreSystem.cs:67)
Microsoft.MixedReality.Toolkit.<>c:<LateUpdateAllServices>b__69_0(IMixedRealityService) (at Library/PackageCache/com.microsoft.mixedreality.toolkit.foundation@2.5.0-preview.20200910.3/Core/Services/MixedRealityToolkit.cs:1035)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit:ExecuteOnAllServicesInOrder(Action`1) (at Library/PackageCache/com.microsoft.mixedreality.toolkit.foundation@2.5.0-preview.20200910.3/Core/Services/MixedRealityToolkit.cs:1109)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit:LateUpdateAllServices() (at Library/PackageCache/com.microsoft.mixedreality.toolkit.foundation@2.5.0-preview.20200910.3/Core/Services/MixedRealityToolkit.cs:1035)
Microsoft.MixedReality.Toolkit.MixedRealityToolkit:LateUpdate() (at Library/PackageCache/com.microsoft.mixedreality.toolkit.foundation@2.5.0-preview.20200910.3/Core/Services/MixedRealityToolkit.cs:700)

```

## Expected behavior

No warnings should appear when playing empty scene configured with MRTK.

## Your setup (please complete the following information)

- Unity 2019.4.10f1 Personal
- MRTK Version 2.5 UPM

## Target platform (please complete the following information)
- Run in Unity Editor, target platform was set to Standalone

**Comment by wiwei** [comment]
Also note, this is happening with .unitypackages as well

**Comment by julenka** [comment]
Marking release blocker based on our policy of not introducing any new warnings when playing using default MRTK profiles.

**Comment by keveleigh** [comment]
@RogPodge Was the HoverLight added as part of your teleport cursor upgrades in https://github.com/microsoft/MixedRealityToolkit-Unity/pull/8386? I don't remember having one on the cursor before

**Comment by wiwei** [comment]
Yeah I was looking too, looks like the HoverLight is getting pulled in when the TeleportPointer starts up with hands (i.e. in editor simulation hands).

**Comment by wiwei** [comment]
We could increase the hover light capacity (though this I think reduces our perf). It's possible that this might be a spurious warning (i.e. if the teleport pointer isn't visible then I assume there's no hover light involved, same if the teleport pointer is causing the other things to not show).

**Comment by RogPodge** [comment]
yeah part of the teleporter pointer improvements involved adding the Hoverlight to better illuminate the teleport area, I'm not sure what the best call to make here is with regards to perf vs. improving the user experience. I'd lean towards it being ok from a performance perspective, since the lights aren't active unless you are teleporting, and upping the count to say 4 should be alright.

**Comment by RogPodge** [comment]
We already seem open to the user manually increasing the number of hoverlights used here, so I feel like it should be ok to bump this number up a bit, thoughts?

https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Rendering/HoverLight.html

**Comment by keveleigh** [comment]
@Cameron-Micka for HoverLight feedback

**Comment by Cameron-Micka** [comment]
This is a warning that more hover lights are active in the scene than the shader will accept. This could be likely spurious as @wiwei mentions if it's impossible to have two hands active while teleporting? 

If this is a feature that is needed, it would align with another request: https://github.com/microsoft/MixedRealityToolkit-Unity/issues/5888 I think I have a solution in mind that would keep performance the same for existing setups and allow users to add more lights if needed. (The only drawback would be more shader permutations, aka increase build times and memory footprint. But, I don't anticipate it would be drastic.)

**Comment by RogPodge** [comment]
@Cameron-Micka is it ok to up the limit to 4 for now so the user doesn't encounter the warnings? This is to unblock the 2.5 release. I don't really want to remove the hover lights from the teleport cursors because who knows if we'll ever get around to remembering to add them back after you make your performance changes.

**Comment by Cameron-Micka** [comment]
Hi @RogPodge, I did a quick test to see how many additional instructions would result from increasing the count from 2 to 4:

```
 // 2 Hover Lights Max
 // Stats for Vertex shader:
 //        d3d11: 23 avg math (8..88)
 // Stats for Fragment shader:
 //        d3d11: 62 avg math (1..145), 0 avg texture (0..1)

 // 4 Hover Lights Max
 // Stats for Vertex shader:
 //        d3d11: 23 avg math (8..100)
 // Stats for Fragment shader:
 //        d3d11: 72 avg math (1..161), 0 avg texture (0..1)
```

You can see the average case introduces 10 extra pixel shader instructions (and 16 in the worst case). It's hard to quantify these numbers since I bet it would introduce little performance loss in our example levels (since many of them are rather simple). But, might have adverse effects in apps which are already close to being GPU bound.

Afaik we don't have any test harnesses to stress test changes like this (especially since our tests don't happen on device). But, they would be nice to have!

How close is the 2.5 release? Or, how soon would the change I describe above need to be committed?

**Comment by RogPodge** [comment]
I believe we're trying to release before the end of the month if I'm not mistaken  @wiwei @Cameron-Micka

**Comment by elbuhofantasma** [comment]
I also ran into this issue when using the Unity in-editor player on the Android platform with the Hand Interaction Examples scene. I had the Default XR SDK Profile enabled when the issue occurred.

![issue8543](https://user-images.githubusercontent.com/65038391/93630972-b6c98d80-f99f-11ea-9e8e-ef5522354bbc.png)

**Comment by Cameron-Micka** [comment]
Just opened a PR to (hopefully) address this: https://github.com/microsoft/MixedRealityToolkit-Unity/pull/8596
